Update Building with Microsoft Visual Studio Instructions

Codym48 opened the pull request:
## Description

Update the **Building with Microsoft Visual Studio** section of the INSTALL instructions.

Add details on required optional packages from Visual Studio 2015 Community install.

Rearrange instructions and provide exact commands for example directory (C:\\OSGeo4W64\\QGIS). The previous instructions used a mix of c:\\, d:\\, D:\\, and X:\\ paths, some of which didn't exist yet at that point in the workflow (point to a QGIS source file before cloning the repository).

~~Add gdal+proj libraries to PATH in MSVC environment setup script. More specifically: call gdal-dev-env.bat as part of msvc-env.bat. The package-nightly.cmd script already called this additional .bat file to set up the PATH to gdal and proj libraries (added in bc1aea62):~~

      call msvc-env.bat %ARCH%
      call gdal-dev-env.bat

~~So it makes sense to move this call into msvc-env.bat?~~

Tell the user to `call gdal-dev-env.bat` in the command prompt before opening MSVC.

Without doing this, a build of the qgis.sln file in MSVC fails. Specifically, the synccrsdb project fails with:

      #>  Running crssync
      #>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code -1073741515.

Running crssync.exe from the command line also fails:

      C:\OSGeo4W64>QGIS>ms-windows\osgeo4w\build-qgis-test-x86_64\output\bin\RelWithDebInfo\crssync.exe
      The code execution cannot proceed because gdal302.dll was not found. Reinstalling the program may fix this problem.
      The code execution cannot proceed because proj_7_1.dll was not found. Reinstalling the program may fix this problem.
      The code execution cannot proceed because gdal302.dll was not found. Reinstalling the program may fix this problem.

After this change, the configonly.bat workflow produces a qgis.sln file that builds to completion.

Following this updated workflow with this branch on Windows 10 works!

![qgisbanner](https://user-images.githubusercontent.com/34431661/81097968-86d09080-8ebd-11ea-949c-59f49e7cf08b.png)
![itworks](https://user-images.githubusercontent.com/34431661/81097981-8afcae00-8ebd-11ea-8e25-b56a7c621687.png)


## Todo

 - [x] Use txt2tags to update docs/INSTALL.html and INSTALL
 - [x] Resolve "Couldn't load plugin 'MetaSearch'" error due to missing owslib

Codym48 committed:
Switch to //Advanced Install// to exactly match installation prompt

Codym48 committed:
Switch from [] to () to prevent creation of broken hyperlinks

Codym48 committed:
Add detail about required packages to VS2015 Community install instructions

Codym48 committed:
Clarify Windows configonly.bat workflow and fix user-specific path

The previous workflow encouraged the user to create a batch file
pointing to a nonexistant file before cloning the source code
repository.

Rearrange the calls so that the user can succesfully clone the QGIS
repository, then create a batch file that calls a file in that
source tree, then open the command prompt created by that batch
file, then call a series of commands that configures using CMake,
builds using Visual Studio, and runs QGIS.

Codym48 committed:
Suggest moving into local QGIS repository clone before executing git config command

Codym48 committed:
Add gdal+proj libraries to PATH in MSVC environment setup script

More specifically: call gdal-dev-env.bat as part of msvc-env.bat.

Without doing this, the a build of the qgis.sln file in MSVC fails.
Specifically, the synccrsdb project fails with:

  #>  Running crssync
  #>C:\Program Files (x86)\MSBuild\Microsoft.Cpp\v4.0\V140\Microsoft.CppCommon.targets(171,5): error MSB6006: "cmd.exe" exited with code -1073741515.

Running crssync.exe from the command line also fails:

  C:\OSGeo4W64>QGIS>ms-windows\osgeo4w\build-qgis-test-x86_64\output\bin\RelWithDebInfo\crssync.exe
  The code execution cannot proceed because gdal302.dll was not found. Reinstalling the program may fix this problem.
  The code execution cannot proceed because proj_7_1.dll was not found. Reinstalling the program may fix this problem.
  The code execution cannot proceed because gdal302.dll was not found. Reinstalling the program may fix this problem.

With this change, the configonly.bat workflow produces a qgis.sln
file that builds to completion.

The package-nightly.cmd script already called this additional .bat
file to set up the PATH to gdal and proj libraries:

  call msvc-env.bat %ARCH%
  call gdal-dev-env.bat

So it makes sense to add this call into msvc-env.bat.

Codym48 committed:
Remove now-redundant gdal environment setup script from package-nightly.cmd

Codym48 committed:
Add some detail to the Run instructions

Codym48 committed:
Clarify sentence about CMake generator options

Codym48 commented on doc/msvc.t2t:
```diff
@@ -7,12 +7,17 @@ This is currently also how the binary QGIS packages are made (earlier versions u
 	http://download.microsoft.com/download/D/2/3/D23F4D0F-BA2D-4600-8725-6CCECEA05196/vs_community_ENU.exe
 
+Select "Custom" install and add the following packages:
+- "Common Tools for Visual C++ 2015" under "Visual C++"
```
Required to avoid:
![visualC++2015tools](https://user-images.githubusercontent.com/34431661/81096064-b92cbe80-8eba-11ea-9b6d-be9a33203a0e.png)

Codym48 commented on doc/msvc.t2t:
```diff
@@ -7,12 +7,17 @@ This is currently also how the binary QGIS packages are made (earlier versions u
 
+Select "Custom" install and add the following packages:
+- "Common Tools for Visual C++ 2015" under "Visual C++"
+- "Tools (1.4.1) and Windows 10 SDK (10.0.14393)" under "Universal Windows App Development Tools".
```
Required to avoid erroring out [here in msvc-env.bat](https://github.com/Codym48/QGIS/blob/838bebea83ac1cb78ca46a2c7a40872bbc54b625/ms-windows/osgeo4w/msvc-env.bat#L42).

Codym48 commented on ms-windows/osgeo4w/msvc-env.bat:
```diff
@@ -55,6 +55,7 @@ if not exist "%OSGEO4W_ROOT%\bin\o4w_env.bat" (echo o4w_env.bat not found & goto
 call "%OSGEO4W_ROOT%\bin\o4w_env.bat"
 call "%OSGEO4W_ROOT%\bin\py3_env.bat"
 call "%OSGEO4W_ROOT%\bin\qt5_env.bat"
+call "%OSGEO4W_ROOT%\bin\gdal-dev-env.bat"
```
grep tells me the only other thing calling this file is [package.cmd](https://github.com/qgis/QGIS/blob/0ada65b210cc57814393d00ec82312e6703940fe/ms-windows/osgeo4w/package.cmd#L37). Will this new line be a problem?

jef-n replied:
yes, the nightlies are build against the proj and nightlies, while the release package are build against the release packages.

Codym48 replied:
Requirements:
 - msvc-env.bat shall not impose specific gdal or proj libraries so that it can be used in package.cmd and package-nightly.cmd
 - package.cmd shall pull from nightly builds of gdal and proj libraries
 - package-nightly.cmd shall pull from release gdal and proj libraries in apps\*-dev\bin

Desirement:
 - user can execute exact **Building with Microsoft Visual Studio** workflow to generate and run a working QGIS executable

Rather than moving the call into msvc-env.bat, would it be better to direct the user to `call gdal-dev-env.bat` in the command prompt?

jef-n commented on doc/msvc.t2t:
```diff
@@ -46,63 +51,83 @@ dependencies.  If you're interested in that, check the history of this page in t
-/!\ Consider this section as example.  It tends to outdate, when OSGeo4W and
-SDKs move on.  ``ms-windows/osgeo4w/package-nightly.cmd`` is used for the
-nightly builds and constantly updated and hence might contain necessary
-updates that are not yet reflected here.
```
IMHO this should stay.

jef-n commented on doc/msvc.t2t:
```diff
@@ -46,63 +51,83 @@ dependencies.  If you're interested in that, check the history of this page in t
-To start a command prompt with an environment that both has the VC++ and the OSGeo4W
-variables create the following batch file (assuming the above packages were
-installed in the default locations):
+Chose a directory to store the QGIS source code.
```
typo

Codym48 replied:
👍🙏

Codym48 committed:
Spell choose correctly

Codym48 committed:
Revert "Add gdal+proj libraries to PATH in MSVC environment setup script"

This reverts commit b8c803fc7e2b7fa958247d27777f6c4d53c4e136.

Codym48 committed:
Revert "Remove now-redundant gdal environment setup script from package-nightly.cmd"

This reverts commit 4032d43884bf5355ac5574f0575e8f539ad86f59.

Codym48 committed:
Suggest that user call gdal-dev-env.bat before opening MSVC

Codym48 committed:
Regenerate INSTALL[.html] with txt2tags 3.6 minus Mac changes

It looks like #35915 and 4f91bee1 regenerated INSTALL from osx.t2t
but didn't touch INSTALL.html. In this commit, I backed out all
changes to the Mac/OSX sections to both INSTALL and INSTALL.html
after using txt2tags to generate the files, to keep those sections
identical to what came before me.

Codym48 committed:
Regenerate INSTALL[.html] with txt2tags 3.6 including Mac changes

From C:\OSGeo4W64 with the 3.6 tag of txt2tags checked out
at C:\OSGeo4W64\tx2tags and this branch of QGIS checked out
at C:\OSGeo4W64\QGIS, call:

python txt2tags\txt2tags.py -oQGIS\INSTALL -t txt QGIS\doc\INSTALL.t2t
python txt2tags\txt2tags.py -oQGIS\doc\INSTALL.html -t html QGIS\doc\INSTALL.t2t

Codym48 committed:
Try to normalize line endings to LF with * text=auto

Codym48 commented:
I think this error at startup is due to not having `owslib` in my OSGeo4W install. Apparently, this package isn't included as a dependency of [the `qgis-dev-deps` package](https://github.com/qgis/QGIS/blob/5bf120a885ff3344ae1724add38f190de152504b/INSTALL#L508)?

```
Couldn't load plugin 'MetaSearch' due to an error when calling its classFactory() method 

ModuleNotFoundError: No module named 'owslib' 
Traceback (most recent call last):
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 334, in _startPlugin
    plugins[packageName] = package.classFactory(iface)
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins\MetaSearch\__init__.py", line 29, in classFactory
    from MetaSearch.plugin import MetaSearchPlugin
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins\MetaSearch\plugin.py", line 34, in 
    from MetaSearch.dialogs.maindialog import MetaSearchDialog
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins\MetaSearch\dialogs\maindialog.py", line 50, in 
    from owslib.csw import CatalogueServiceWeb # spellok
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
ModuleNotFoundError: No module named 'owslib'


Python version: 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)] 
QGIS version: 3.13.0-Master Master, 838bebea83 

Python Path:
C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python
C:/Users/codyt/AppData/Roaming/QGIS/QGIS3\profiles\default/python
C:/Users/codyt/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins
C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins
C:\OSGeo4W64\apps\Python37\python37.zip
C:\OSGeo4W64\apps\Python37\DLLs
C:\OSGeo4W64\apps\Python37\lib
C:\OSGeo4W64\QGIS\ms-windows\osgeo4w\build-qgis-test-x86_64\output\bin\RelWithDebInfo
C:\OSGeo4W64\apps\Python37
C:\OSGeo4W64\apps\Python37\lib\site-packages
C:\OSGeo4W64\apps\Python37\lib\site-packages\win32
C:\OSGeo4W64\apps\Python37\lib\site-packages\win32\lib
C:\OSGeo4W64\apps\Python37\lib\site-packages\Pythonwin
C:/Users/codyt/AppData/Roaming/QGIS/QGIS3\profiles\default/python
```

It doesn't prevent me from using QGIS to do the training exercises, so maybe leaving it out is intentional.

Should we be telling users to install the `owslib` and `python3-owslib` libraries as part of their custom OSGeo4W install? We [warn them against doing it right now](https://github.com/qgis/QGIS/blob/5bf120a885ff3344ae1724add38f190de152504b/INSTALL#L512).

Codym48 commented on INSTALL:
```diff
@@ -501,9 +506,9 @@ For the QGIS build you need to install following packages from cygwin:
-and from OSGeo4W (select Advanced Installation):
+and from OSGeo4W (select Advanced Install):
 
 - qgis-dev-deps
```
I think this error at startup is due to not having `owslib` in my OSGeo4W install. Apparently, this package isn't included as a dependency of [the `qgis-dev-deps` package](https://github.com/qgis/QGIS/blob/5bf120a885ff3344ae1724add38f190de152504b/INSTALL#L508)?

```
Couldn't load plugin 'MetaSearch' due to an error when calling its classFactory() method 

ModuleNotFoundError: No module named 'owslib' 
Traceback (most recent call last):
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 334, in _startPlugin
    plugins[packageName] = package.classFactory(iface)
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins\MetaSearch\__init__.py", line 29, in classFactory
    from MetaSearch.plugin import MetaSearchPlugin
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins\MetaSearch\plugin.py", line 34, in 
    from MetaSearch.dialogs.maindialog import MetaSearchDialog
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins\MetaSearch\dialogs\maindialog.py", line 50, in 
    from owslib.csw import CatalogueServiceWeb # spellok
  File "C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python\qgis\utils.py", line 743, in _import
    mod = _builtin_import(name, globals, locals, fromlist, level)
ModuleNotFoundError: No module named 'owslib'


Python version: 3.7.0 (v3.7.0:1bf9cc5093, Jun 27 2018, 04:59:51) [MSC v.1914 64 bit (AMD64)] 
QGIS version: 3.13.0-Master Master, 838bebea83 

Python Path:
C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python
C:/Users/codyt/AppData/Roaming/QGIS/QGIS3\profiles\default/python
C:/Users/codyt/AppData/Roaming/QGIS/QGIS3\profiles\default/python/plugins
C:/OSGeo4W64/QGIS/ms-windows/osgeo4w/build-qgis-test-x86_64/output/python/plugins
C:\OSGeo4W64\apps\Python37\python37.zip
C:\OSGeo4W64\apps\Python37\DLLs
C:\OSGeo4W64\apps\Python37\lib
C:\OSGeo4W64\QGIS\ms-windows\osgeo4w\build-qgis-test-x86_64\output\bin\RelWithDebInfo
C:\OSGeo4W64\apps\Python37
C:\OSGeo4W64\apps\Python37\lib\site-packages
C:\OSGeo4W64\apps\Python37\lib\site-packages\win32
C:\OSGeo4W64\apps\Python37\lib\site-packages\win32\lib
C:\OSGeo4W64\apps\Python37\lib\site-packages\Pythonwin
C:/Users/codyt/AppData/Roaming/QGIS/QGIS3\profiles\default/python
```

It doesn't prevent me from using QGIS to do the training exercises, so maybe leaving it out is intentional.

Should we be telling users to install the `owslib` and `python3-owslib` libraries as part of their custom OSGeo4W install? We [warn them against doing it right now](https://github.com/qgis/QGIS/blob/5bf120a885ff3344ae1724add38f190de152504b/INSTALL#L512).

jef-n replied:
> It doesn't prevent me from using QGIS to do the training exercises, so maybe leaving it out is intentional.

It's not a build dependency.

Codym48 replied:
> > It doesn't prevent me from using QGIS to do the training exercises, so maybe leaving it out is intentional.
> 
> It's not a build dependency.

👍 then I think this is ready to merge.
